### ISSUE TITLE
DNS based service discovery 

### ISSUE THREAD

**Issue by luisdavim** [buraksezer/olric#224]
Hi, this relates to https://github.com/buraksezer/olric/issues/195 but it's more of a feature request. 

I know tool/library supports different service discovery mechanisms through plugins but it would be nice to have one working out of the box without needing any plugins.

I think DNS would be perfect for that, it can be used with `consul`, `kubernetes` and any cloud provider and since the implementation can be done using the standard library, it doesn't require any additional dependencies.

**Comment by buraksezer** [comment]
Hey @luisdavim,

Thank you for reporting this. It would be good having a default implementation based on DNS. Olric already depends on (indirectly) `github.com/miekg/dns`. I think this task can be a good first issue. Please feel free to provide a design document and create a PR. for this feature.

**Comment by kozlowsqi** [comment]
Is this still relevant?

I am looking for my first issue and this seems easy enough.

**Comment by luisdavim** [comment]
Yes, I think this is still relevant.

**Comment by buraksezer** [comment]
> Is this still relevant?
> 
> I am looking for my first issue and this seems easy enough.

Hello, 

Yes, it's still relevant but we should start working on an RFC before implementing the feature.

**Comment by kozlowsqi** [comment]
Ok. I am sorry but I am not sure what you mean by an RFC. 

If I understood this issue correctly, what you want is something similar to how [nginx dns service discovery](https://www.nginx.com/blog/dns-service-discovery-nginx-plus/) works.

My initial idea is to implement the `ServiceDiscovery` interface and simply create a default plugin for service discovery. 

However I am quite green and would appreciate some guidance.

**Comment by luisdavim** [comment]
Plugins for service discovery already exist. The point of this issue is to have a native, default service discovery mechanism.

**Comment by kozlowsqi** [comment]
> Plugins for service discovery already exist. The point of this issue is to have a native, default service discovery mechanism.

If I understand it correctly, plugins for service discovery are written in Go, and they implement an interface from [here](https://github.com/buraksezer/olric/blob/7cd094082f2027ee556f0e9c2007fafe16081f2c/pkg/service_discovery/service_discovery.go#L21C16-L21C16). That package is the common dependency between orlic and service discovery plugin (orlic consumes the interface, the plugin implements it).

All I would need to do is provide a default implementation that is used instead where appropriate. Am I on the right track here?

**Comment by luisdavim** [comment]
The idea is that we would have a default service discovery mechanism based on DNS, without having to use plugins. 

An RFC (request for comments) is a document where you would explain the proposed implementation so it cabeceiras reviewed and agreed upon before starting the actual implementation.

**Comment by Tochemey** [comment]
I have a sort of implementation in this library here https://github.com/Tochemey/goakt/blob/main/discovery/dnssd/discovery.go. I could export it if it does fit the expectation of this task.
